cdap-app-templates/cdap-etl/cdap-calcite/src/main/java/io/cdap/cdap/datapipeline/relational/ExpressionConversionContext.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.datapipeline.relational;
+
+import io.cdap.cdap.api.data.schema.Schema;
+
+import javax.annotation.Nullable;
+
+/**
+ * A class that provides a Context object to facilitate working with {@link SQLDialectConversionExpressionFactory}
+ * objects with specific SQL dialects
+ */
+public class ExpressionConversionContext {
+
+    SQLDialectConverter converter;
+
+    public ExpressionConversionContext() {
+        converter = new CalciteSQLDialectConverter();
+    }
+
+    /**
+     * Returns a conversion factory initialized with the source and destination SQL dialects specified and
+     * an optional schema
+     * @param src The SQL dialect in which expressions will be specified
+     * @param dest The SQL dialect in which expressions are to be output after conversion
+     * @param schema The schema of the table on which the SQL operation is performed
+     * @return An expression factory that will accept SQL expressions in the source dialect and convert them into
+     * expressions in the destination dialect after validation, taking the schema into account
+     */
+    public SQLDialectConversionExpressionFactory getConversionFactory(
+            SQLDialect src, SQLDialect dest, @Nullable Schema schema) {
```

Comment:
Who is going to call this and how? With current DL design you don't have schema when you are creating factory. In our discussions I proposed some ways to improve it. Let's define the interface and the usage. We can even separate implementation into separate PR to make it easier

Reply:
This will be called by the plugin (Wrangler). It will use the context to get the expression factory and then call `compile()` which will also call `validate()`.

Reply:
Neither io.cdap.cdap.etl.api.relational.Engine#getExpressionFactories nor io.cdap.cdap.etl.api.relational.ExpressionFactory#compile take schema. If you plan to change the API, this should be included in this PR or better in a separate PR before this one. Don't create implementation (how) before API (what)

Reply:
@tivv Do you have any suggestions on how to structure this? The schema is required to be able to validate that the expression makes sense and is  valid (e.g. not referencing invalid column names)

Reply:
I had suggestions and I can bring them, but this PR is a wrong place. We need a first PR with the flow (API / API calls) and mock implementation and this should be the second PR that adds the meat, not visa versa

Reply:
As per the current implementation, the schema will be passed to `SQLDialectConversionExpressionFactory` through a constructor. None of the existing methods of `Engine` or `ExpressionFactory` need to be changed, so there's no API change.